File: composeApp/src/commonMain/kotlin/com/cochelper/app/domain/StImport.kt
package com.cochelper.app.domain

import com.cochelper.app.data.local.SkillItem

/**
 * `.st` 骰娘指令解析器：把一段一次性粘贴的指令解析为「属性 + 技能」两部分。
 *
 * 兼容常见写法：
 *  - `.st 力量50敏捷60`（名称数值无空格直接拼接）
 *  - `.st 力量 50 敏捷 60` / `.st 力量:50` / `力量=50`
 *  - 条目间用 `|` / `,` / `、` / `;` 分隔
 *  - `名字--力量30`（带角色卡名前缀）
 *  - 数值支持整数或骰式 `3d6`、`3d6*5`、`3d6+5`
 *  - 属性名中英文均识别（力量/STR、敏捷/DEX、意志/POW…）
 */
object StImport {

    data class Result(
        /** 规范中文属性名 → 数值（如 "力量" → 50），已含派生属性（生命/理智/魔法）。 */
        val attributes: Map<String, Int>,
        val skills: List<SkillItem>,
    )

    /** 中文 + 英文属性名 → 规范中文名（键统一转大写后匹配）。 */
    private val attributeAliases: Map<String, String> = buildMap {
        listOf(
            "生命" to "生命", "理智" to "理智", "魔法" to "魔法", "移动力" to "移动力",
            "力量" to "力量", "敏捷" to "敏捷", "意志" to "意志", "体质" to "体质",
            "外貌" to "外貌", "教育" to "教育", "体型" to "体型", "智力" to "智力", "幸运" to "幸运",
            "HP" to "生命", "SAN" to "理智", "MP" to "魔法", "MOV" to "移动力",
            "STR" to "力量", "DEX" to "敏捷", "POW" to "意志", "CON" to "体质",
            "APP" to "外貌", "EDU" to "教育", "SIZ" to "体型", "INT" to "智力", "LUCK" to "幸运",
        ).forEach { (alias, canonical) -> put(alias.uppercase(), canonical) }
    }

    // 名称（非数字、非空白）+ 数值（整数，或 XdY 骰式可带 *k / +k / -k）
    private val pairRegex = Regex("""([^\d\s]+?)(\d+[dD]\d+(?:[*+\-]\d+)?|\d+)""")

    fun parse(raw: String): Result {
        var text = raw.trim()

        // 1) 去掉指令前缀 .st / /st / !st
        val lower = text.lowercase()
        text = when {
            lower.startsWith(".st") || lower.startsWith("/st") || lower.startsWith("!st") -> text.drop(3)
            else -> text
        }

        // 2) 去掉角色卡名前缀：名字--属性
        val dd = text.indexOf("--")
        if (dd >= 0) text = text.substring(dd + 2)

        // 3) 归一化分隔符：| , 、 ; 分隔条目；: = 分隔名称与数值；最后移除所有空白
        text = text
            .replace('|', ' ').replace(',', ' ').replace('，', ' ').replace('、', ' ')
            .replace(';', ' ').replace('；', ' ')
            .replace(':', ' ').replace('：', ' ').replace('=', ' ')
            .replace(Regex("""\s+"""), "")

        val attributes = mutableMapOf<String, Int>()
        val skillsByName = linkedMapOf<String, Int>()

        pairRegex.findAll(text).forEach { m ->
            val name = m.groupValues[1].trimEnd('+', '-', '*', '(', ')')
            val value = evaluateValue(m.groupValues[2])
            if (name.isEmpty() || value == null) return@forEach
            val canonical = attributeAliases[name.uppercase()]
            if (canonical != null) {
                attributes[canonical] = value
            } else {
                skillsByName[name] = value
            }
        }

        // 派生属性：仅当用户未显式给出时按公式补全。
        val pow = attributes["意志"] ?: 0
        val con = attributes["体质"] ?: 0
        val siz = attributes["体型"] ?: 0
        if ("生命" !in attributes && (con > 0 || siz > 0)) attributes["生命"] = (con + siz) / 10
        if ("理智" !in attributes && pow > 0) attributes["理智"] = pow
        if ("魔法" !in attributes && pow > 0) attributes["魔法"] = pow / 5

        return Result(attributes, skillsByName.map { (n, v) -> SkillItem(n, v) })
    }

    /** 解析数值：整数直接返回；`XdY` 掷骰求和（可带 *k / +k / -k）后返回；无法解析返回 null。 */
    private fun evaluateValue(s: String): Int? {
        s.toIntOrNull()?.let { return it }
        val m = Regex("""^(\d+)[dD](\d+)(?:([*+\-])(\d+))?$""").matchEntire(s) ?: return null
        val count = m.groupValues[1].toInt().coerceIn(1, 100)
        val faces = m.groupValues[2].toInt().coerceIn(1, 10000)
        var total = DiceEngine.rollDice(faces, count).sum()
        val op = m.groupValues[3]
        val k = m.groupValues[4].toIntOrNull()
        if (op.isNotEmpty() && k != null) {
            total = when (op) {
                "*" -> total * k
                "+" -> total + k
                "-" -> total - k
                else -> total
            }
        }
        return total
    }
}

File: web/src/lib/domain/stImport.ts
import { rollDice } from './dice';
import type { SkillItem } from '$lib/data/types';

export interface StResult {
  /** 规范中文属性名 → 数值（已含派生属性：生命/理智/魔法）。 */
  attributes: Record<string, number>;
  skills: SkillItem[];
}

const ALIASES: [string, string][] = [
  ['生命', '生命'], ['理智', '理智'], ['魔法', '魔法'], ['移动力', '移动力'],
  ['力量', '力量'], ['敏捷', '敏捷'], ['意志', '意志'], ['体质', '体质'],
  ['外貌', '外貌'], ['教育', '教育'], ['体型', '体型'], ['智力', '智力'], ['幸运', '幸运'],
  ['HP', '生命'], ['SAN', '理智'], ['MP', '魔法'], ['MOV', '移动力'],
  ['STR', '力量'], ['DEX', '敏捷'], ['POW', '意志'], ['CON', '体质'],
  ['APP', '外貌'], ['EDU', '教育'], ['SIZ', '体型'], ['INT', '智力'], ['LUCK', '幸运']
];

const attributeAliases = new Map<string, string>(
  ALIASES.map(([alias, canonical]) => [alias.toUpperCase(), canonical])
);

// 名称（非数字、非空白）+ 数值（整数，或 XdY 骰式可带 *k / +k / -k）
const pairRegex = /([^\d\s]+?)(\d+[dD]\d+(?:[*+\-]\d+)?|\d+)/g;

function evaluateValue(s: string): number | null {
  if (/^\d+$/.test(s)) return parseInt(s, 10);
  const m = /^(\d+)[dD](\d+)(?:([*+\-])(\d+))?$/.exec(s);
  if (!m) return null;
  const count = Math.min(Math.max(parseInt(m[1], 10), 1), 100);
  const faces = Math.min(Math.max(parseInt(m[2], 10), 1), 10000);
  let total = rollDice(faces, count).reduce((a, b) => a + b, 0);
  const op = m[3] ?? '';
  const k = m[4] != null ? parseInt(m[4], 10) : null;
  if (op && k != null) {
    if (op === '*') total = total * k;
    else if (op === '+') total = total + k;
    else if (op === '-') total = total - k;
  }
  return total;
}

export function parseSt(raw: string): StResult {
  let text = raw.trim();

  // 1) 去掉指令前缀 .st / /st / !st
  const lower = text.toLowerCase();
  if (lower.startsWith('.st') || lower.startsWith('/st') || lower.startsWith('!st')) {
    text = text.slice(3);
  }

  // 2) 去掉角色卡名前缀：名字--属性
  const dd = text.indexOf('--');
  if (dd >= 0) text = text.slice(dd + 2);

  // 3) 归一化分隔符
  text = text
    .replace(/\|/g, ' ').replace(/,/g, ' ').replace(/，/g, ' ').replace(/、/g, ' ')
    .replace(/;/g, ' ').replace(/；/g, ' ')
    .replace(/:/g, ' ').replace(/：/g, ' ').replace(/=/g, ' ')
    .replace(/\s+/g, '');

  const attributes: Record<string, number> = {};
  const skillsByName = new Map<string, number>();

  for (const m of text.matchAll(pairRegex)) {
    const name = m[1].replace(/[-+*()]+$/, '');
    const value = evaluateValue(m[2]);
    if (!name || value == null) continue;
    const canonical = attributeAliases.get(name.toUpperCase());
    if (canonical != null) {
      attributes[canonical] = value;
    } else {
      skillsByName.set(name, value);
    }
  }

  // 派生属性：仅当用户未显式给出时按公式补全。
  const pow = attributes['意志'] ?? 0;
  const con = attributes['体质'] ?? 0;
  const siz = attributes['体型'] ?? 0;
  if (!('生命' in attributes) && (con > 0 || siz > 0)) attributes['生命'] = Math.floor((con + siz) / 10);
  if (!('理智' in attributes) && pow > 0) attributes['理智'] = pow;
  if (!('魔法' in attributes) && pow > 0) attributes['魔法'] = Math.floor(pow / 5);

  return {
    attributes,
    skills: [...skillsByName].map(([name, value]) => ({ name, value }))
  };
}
